### PR TITLE
Add a button to reset personal encryption state during login

### PR DESCRIPTION
**Pull request by turt2live**
*No public issue.*

----

**Note**: This deliberately gets code review first as changes might affect how the design review goes. This needs design review to land.

Also fair warning I copy/pasted most of this in an effort to reuse existing flows.

----

image
image

... leads to:
image

### CODE REVIEW

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx [hunk 1]
```diff
@@ -201,6 +208,50 @@ export default class AccessSecretStorageDialog extends React.PureComponent<IProp
...
+        // rest of this function.
+        Modal.toggleCurrentDialogVisibility();
+
+        // Force reset secret storage (which resets the key backup)
+        await accessSecretStorage(async () => {
+            // Now reset cross-signing so everything Just Works™ again.
+            const cli = MatrixClientPeg.get();
+            await cli.bootstrapCrossSigning({
+                authUploadDeviceSigningKeys: async (makeRequest) => {
+                    // XXX: Making this an import breaks the app.
```
Comment (jryans) [inline src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:231]: Hmm, it works fine in other components... 🤔

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx [hunk 2]
```diff
@@ -201,6 +208,50 @@ export default class AccessSecretStorageDialog extends React.PureComponent<IProp
...
+                    if (!confirmed) {
+                        throw new Error("Cross-signing key upload auth canceled");
+                    }
+                },
+                setupNewCrossSigning: true,
+            });
+
+            // Now we can indicate that the user is done pressing buttons, finally.
+            // Upstream flows will detect the new secret storage, key backup, etc and use it.
+            this.props.onFinished(true);
```
Comment (jryans) [inline src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:251]: What happens if something above throws an error...? We won't get here, so will the UI appear to hang?

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx [hunk 3]
```diff
@@ -226,11 +278,36 @@ export default class AccessSecretStorageDialog extends React.PureComponent<IProp
...
         let title;
         let titleClass;
-        if (hasPassphrase && !this.state.forceRecoveryKey) {
-            const DialogButtons = sdk.getComponent('views.elements.DialogButtons');
+        if (this.state.resetting) {
+            title = _t("Reset everything");
+            titleClass = ['mx_AccessSecretStorageDialog_titleWithIcon mx_AccessSecretStorageDialog_resetBadge'];
+            content = <div>
+                <p>{_t("Only do this if you have no other device to complete verification with.")}</p>
+                <p>{_t("If you reset everything, you will restart with no trusted devices, no trusted users, and "
```
Comment (jryans) [inline src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:299]:
```suggestion
                <p>{_t("If you reset everything, you will restart with no trusted sessions, no trusted users, and "
```

**Review by jryans — APPROVED** [review]
Thanks, this overall looks good code-wise, and should help more people get out of tricky spots on their own. 😄

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx (hunk 1 above)
Comment (turt2live) [reply src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:231]: This dialog is loaded async for some reason, which makes it special

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx (hunk 2 above)
Comment (turt2live) [reply src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:251]: almost certainly, yea. Would it be okay to do an `onFinished(false)` in a catch?

Not sure how easily we can recover from errors here.

Comment (jryans) [reply src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:251]: Yeah, I think that's a good place to start at least, it should be enough to percolate up to higher level paths.

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx [hunk 4]
```diff
@@ -222,34 +222,44 @@ export default class AccessSecretStorageDialog extends React.PureComponent<IProp
...
-                    const { finished } = Modal.createTrackedDialog(
-                        'Cross-signing keys dialog', '', InteractiveAuthDialog,
-                        {
-                            title: _t("Setting up keys"),
-                            matrixClient: cli,
-                            makeRequest,
+        try {
+            // Force reset secret storage (which resets the key backup)
+            await accessSecretStorage(async () => {
+                try {
```
Comment (jryans) [inline src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:228]: Why is this inner one needed...? `accessSecretStorage` re-throws an errors that happen inside it, so they should all bubble up to your outer catch...

**Review by jryans — CHANGES_REQUESTED** [review]

**Review by gaelledel — APPROVED** [review]

File: src/components/views/dialogs/security/AccessSecretStorageDialog.tsx (hunk 4 above)
Comment (turt2live) [reply src/components/views/dialogs/security/AccessSecretStorageDialog.tsx:228]: Oh, does it? Neat :D

**Review by jryans — APPROVED** [review]

**Pull request merged**